kernel/sched: Add "thread_usage" API for thread runtime cycle monitoring

andyross opened the pull request:
This is an alternate backend that does what THREAD_RUNTIME_STATS is doing currently, but with a few advantages:

* Correctly synchronized: you can't race against a running thread (potentially on another CPU!) while querying its usage.

* Realtime results: you get the right answer always, up to timer precision, even if a thread has been running for a while uninterrupted and hasn't updated its total.

* Portable, no need for per-architecture code at all for the simple case. (It leverages the USE_SWITCH layer to do this, so won't work on older architectures)

* Faster/smaller: minimizes use of 64 bit math; lower overhead in thread struct (keeps the scratch "started" time in the CPU struct instead).  One 64 bit counter per thread and a 32 bit scratch register in the CPU struct.

* Standalone.  It's a core (but optional) scheduler feature, no dependence on para-kernel configuration like the tracing infrastructure.

* More precise: allows architectures to optionally call a trivial zero-argument/no-result cdecl function out of interrupt entry to avoid accounting for ISR runtime in thread totals.  No configuration needed here, if it's called then you get proper ISR accounting, and if not you don't.

For right now, pending unification, it's added side-by-side with the older API and left as a z_*() internal symbol.

Commit:
kernel/sched: Add "thread_usage" API for thread runtime cycle monitoring

This is an alternate backend that does what THREAD_RUNTIME_STATS is
doing currently, but with a few advantages:

* Correctly synchronized: you can't race against a running thread
  (potentially on another CPU!) while querying its usage.

* Realtime results: you get the right answer always, up to timer
  precision, even if a thread has been running for a while
  uninterrupted and hasn't updated its total.

* Portable, no need for per-architecture code at all for the simple
  case. (It leverages the USE_SWITCH layer to do this, so won't work
  on older architectures)

* Faster/smaller: minimizes use of 64 bit math; lower overhead in
  thread struct (keeps the scratch "started" time in the CPU struct
  instead).  One 64 bit counter per thread and a 32 bit scratch
  register in the CPU struct.

* Standalone.  It's a core (but optional) scheduler feature, no
  dependence on para-kernel configuration like the tracing
  infrastructure.

* More precise: allows architectures to optionally call a trivial
  zero-argument/no-result cdecl function out of interrupt entry to
  avoid accounting for ISR runtime in thread totals.  No configuration
  needed here, if it's called then you get proper ISR accounting, and
  if not you don't.

For right now, pending unification, it's added side-by-side with the
older API and left as a z_*() internal symbol.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/xtensa: Add hook for CONFIG_SCHED_THREAD_USAGE accounting in ISRs

Call into z_thread_usage_stop() before ISR entry to avoid including
interrupt handling totals in thread usage stats.

Note that this hook is after the register save and stack swap has
happened, so it still incldues some overhead.  But calling out from
the interrupted stack on Xtensa gets really, really hairy due to the
weird intermediate states we leverage (once we've saved enough context
to make a C call safely, we've lost the ability to use register
windows per the C ABI!).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Currently trying to decide whether it's easier to write a separate test for this feature or add a shim layer to integrate it with RUNTIME_STATS right now and leverage the existing tests.

andyross commented:
And to be clear: notes on functionality differences with CONFIG_RUNTIME_STATS:

* This uses k_cycle_get_32() as the only clock, where the existing feature also hooks timing.h to allow the board/soc/arch layers to insert their own "timing" clocks (I'm on record as being broadly opposed to that subsystem; we should just decide on one "high frequency timing source" defined by the platform, exposed as a single API, and not play with all these indirection games).
* This requires USE_SWITCH to be enabled, so there are architectures where RUNTIME_STATS is the only option

nashif reviewed (changes requested):
> And to be clear: notes on functionality differences with CONFIG_RUNTIME_STATS:
> 
> * This uses k_cycle_get_32() as the only clock, where the existing feature also hooks timing.h to allow the board/soc/arch layers to insert their own "timing" clocks (I'm on record as being broadly opposed to that subsystem; we should just decide on one "high frequency timing source" defined by the platform, exposed as a single API, and not play with all these indirection games).
> * This requires USE_SWITCH to be enabled, so there are architectures where RUNTIME_STATS is the only option

Those two items are essential in the current implementation to support all architectures and the different ways we measure time on different architectures, SoCs and boards.
We need just one way for doing this, ideally building on top of the existing one and integrating all the optimizations this PR is bringing.

Commit:
sched: Hook SCHED_USAGE from existing tracing hook

On older architectures, we don't have the
architecture-independent/scheduler-internal hooks (which require
USE_SWITCH) but there is a hook shared by the tracing layer we can use.

This is sort of a layering violation (stat tracking is a core feature,
tracing is supposed to be optional), but simple and lightweight.  And
eventually it will go away as these architectures migrate.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Add an optional "all" counter for thread_usage

Tally the runtime of all non-idle threads.  Make it optional via
kconfig to avoid overhead.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Add timing.h support to thread_usage

The runtime stats feature has always supported this, so use the same
kconfig to indirect the timing source in the same way.

(Personally I'm not a fan of the "timing" API, which really doesn't do
anything that the existing core "cycles" API does not except add a
bunch of code due to the separate implementation of frequency
management and conversion routines.  It comes from an era where
"cycles" were fixed to a MHz frequency clock on platforms like x86 yet
we had benchmarks that wanted to use the TSC.  Those days are behind
us and "cycles" can be fast everywhere.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Swap RUNTIME_STATS implementation

Clean up RUNTIME_STATS to separate the API from the individual data
backends.  Use the SCHED_THREAD_USAGE tracking instead of the original
for execution_cycles.  Move the kconfig for that into the runtime
stats menu, since it's part of the family now.

Also remove a lot of needless #if's around the declarations.  Unused
structs and uncalled functions don't need to be explicitly hidden.  An
attempt to access a non-existent field (e.g. "execution_cycles" if
that isn't configured) provides all the build time validation we need.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/thread_apis: Fix whitebox assumptions in runtime_stats test

The thread_apis tests of the RUNTIME_STATS feature weren't really
testing the right behavior.

+ It assumed that accounting would only happen at context switch time
  and required that the returned values not change for running threads
  (even CLEARLY running threads like _current!).  But that's not a
  documented feature!  It's actually sort of a wart that we'd like to
  be able to fix (and have fixed, the new backend returns realtime
  values so you can track CPU-bound processes on another CPU).

+ It assumed that k_thread_runtime_stats_all_get() would return time
  that includes idle time (or conversely it forgot that
  k_thread_foreach enumerates over idle threads).  This was sort of a
  bug in the original (because it means that the result is always the
  system uptime multiplied by the number of CPUs)

Broadly, instead of testing the result of a "time" function for
equality (never a good idea) test it via appropriate bounds given the
usage.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Update to unify with the earlier RUNTIME_STATS API, using the new version as the backend.

See comments in the final patch, there were two buglets in the original that are fixed now, but they turned out to violate the whiteboxed test:

* The returned usage is realtime now and will reflect instantaneous state for a running thread.  You can't compare the results for equality unless you know a thread is blocked.
* The "all" variant of the API would include runtime for the system idle threads.  But... that has to be wrong, because that means it's just returning system uptime (multiplied by the number of CPUs).  There is always a thread running!  My assumption is that this was intended to reflect something like Unix "top", which separates out idle time.  So that's what the new version does; it reports CPU used by non-idle threads only.

andyross commented:
Rebased.  Note the typedef warning is a false positive due to code motion.  It's not a new typedef with this PR.

nashif commented:
run 
`west build -b qemu_x86 samples/subsys/shell/shell_module/ -t run`

note the `Total execution cycles: 2504748010 (12473 %)` at the end.

```
uart:~$ kernel threads
Scheduler: 465 since last call
Threads:
 0x113ee0 sysworkq
	options: 0x0, priority: -1 timeout: 0
	state: pending, entry: 0x10bf6b
	Total execution cycles: 1140026 (2 %)
	stack size 1024, unused 700, usage 324 / 1024 (31 %)

*0x112400 shell_uart
	options: 0x0, priority: 14 timeout: 0
	state: queued, entry: 0x106908
	Total execution cycles: 46605132 (87 %)
	stack size 2048, unused 924, usage 1124 / 2048 (54 %)

 0x1122a0 logging
	options: 0x0, priority: 14 timeout: 0
	state: pending, entry: 0x1034cf
	Total execution cycles: 680236 (1 %)
	stack size 768, unused 496, usage 272 / 768 (35 %)

 0x112860 idle 00
	options: 0x1, priority: 15 timeout: 0
	state: , entry: 0x10bae3
	Total execution cycles: 2504748010 (12473 %)
	stack size 320, unused 200, usage 120 / 320 (37 %)

```

andyross commented:
That's reflecting what IMHO was a bug in the original API.  It would report idle thread time in the total reported for "all" threads.  But that's silly, because the result is then just the same as system uptime (well, multiplied by the number of CPUs).  I'm reasonably certain the design intent was to measure the total amount of system runtime consumed by app threads, so that's what the new code does.

Anyway this is now showing the amount of CPU consumed by the idle thread (almost everything) as a fraction of the time the system was running other threads (very little), so the number doesn't make sense.  I'll look at adjusting the sample.  We could add idle thread time to the total returned and report both fractions I suppose.

Commit:
kernel/sched: Separate idle from app thread stats in THREAD_USAGE

It turns out that we have a sample (though not a test) that really
does want to use "k_thread_runtime_stats_all_get()" to measure system
uptime.

Instead of breaking this needlessly, separate the accounting for idle
and non-idle threads.  The legacy API can report their sum, and the
more useful value is available via the kernel struct for future
analysis.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Actually no.  No point in having a fight over design intent.  I added a patch that restores the original sense of k_runtime_stats_all_get() (i.e. returning a system uptime proxy in "execution_cycles"), but left the separate totals for idle and non-idle threads in the kernel struct for future APIs that want them.

nashif commented on arch/xtensa/core/xtensa-asm2.c:
```diff
@@ -147,6 +154,7 @@ static inline unsigned int get_bits(int offset, int num_bits, unsigned int val)
 __unused void *xtensa_int##l##_c(void *interrupted_stack)	\
 {							   \
 	uint32_t irqs, intenable, m;			   \
+	usage_stop();					   \
```
why are we doing this for xtensa only? what about other architectures?

andyross replied:
It's an optimization.  It could absolutely be done for any architecture, it just hasn't been done in this PR.  Again, the current implementation doesn't do this at all.

nashif replied:
> Again, the current implementation doesn't do this at all.

right, and the purpose of this PR is to add this optimization, and it needs to be done for all architectures.

andyross replied:
Again, I don't understand the depth of the criticism here.  This PR does lots of good stuff not related to ISR fidelity in measurements.  And we have optional optimizations everywhere in the architecture API (atomics, switch-vs-swap, sched IPIs, stuff like that).  And this optimization is limited to USE_SWITCH architectures anyway, we can't do it everywhere because there's no portable ISR exit hook on the legacy scheme.

I'll see about adding the architectures I know well, but demanding that new features not be merged except by someone willing to hook the ISR entry code of every architecture in the tree doesn't seem like it's going to scale.

nashif reviewed (commented):
The first commit message is not accurate, talking about adding a parallel implementation, this is not the case anymore.

andyross commented:
FWIW: the first commit message is actually correct, at that point in the tree it's parallel and both backends exist with separate APIs, it then gets inserted as the sole implementation of RUNTIME_STATS in a later commit.  I left it this way for easier review of questions like that, but I can go back and squash I guess.  Honestly this doesn't seem that bad; a little design archaeology in the commit log makes questions about "why was this done?" easier to answer.

nashif commented:
> I left it this way for easier review of questions like that, but I can go back and squash I guess. Honestly this doesn't seem that bad; a little design archaeology in the commit log makes questions about "why was this done?" easier to answer.

Why would we want to introduce one design in one commit to remove it later in the same PR? I do not see how this makes reviews better. A reviewer might go commit by commit and would stop at the first commit asking those questions. We ideally want to have PRs be 'final' and not the work in progress of fixing issues and changing behaviour in followup commits.

andyross commented:
The design isn't removed, nothing is being modified in the code in the original commit here.  We're adding a new utility with its own API in one commit, and then in later commits adding a few features to it to match what previous code did, and then finally inserting it as the backend to the pre-existing API.  I guess I really don't understand the criticism here.  Obviously I can squash this, I just really don't understand how that helps anyone.

Commit:
arch/x86_64: Add hook for CONFIG_SCHED_THREAD_USAGE accounting in ISRs

Call into z_thread_usage_stop() before ISR entry to avoid including
interrupt handling totals in thread usage stats.

This has to go into the assembly immediately before the callback-based
dispatch.  Note that the dispatch code was putting the vector number
in RCX, which was unfortunate as that's a caller-saved register.
Would be nice to clean this up in the future so it lives in a
preserved register but it's mildly complicated to make work with the
way we do the stack layout right now.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/arm64: Add hook for CONFIG_SCHED_THREAD_USAGE accounting in ISRs

Call into z_thread_usage_stop() before ISR entry to avoid including
interrupt handling totals in thread usage stats.

This is pretty much exactly where we want it, just after the context
saving steps (which we can't elide since the hook is in C) and
immediately before the tracing hook for ISR entry.  And as I'm reading
the code, this is purely for Zephyr-registered interrupts, meaning
that software exceptions will be accounted for (correctly) as part of
the excepting thread.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/arc: Add hook for CONFIG_SCHED_THREAD_USAGE accounting in ISRs

Call into z_thread_usage_stop() before ISR entry to avoid including
interrupt handling totals in thread usage stats.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/sparc: Add hook for CONFIG_SCHED_THREAD_USAGE accounting in ISRs

Call into z_thread_usage_stop() before ISR entry to avoid including
interrupt handling totals in thread usage stats.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

carlocaione reviewed (commented):
LGTM for the ARM64 change

andyross commented:
Indeed, there was a edge case in the "obvious" assembly that moved the vector to %rbx.  The entry for nested interrupts makes assumptions about where %rcx lives on the stack that broke.

Which points out that, if the ipm (!) test is the only exerciser we have for nested interrupts on x86_64, then we probably need to write a test or two...
